### PR TITLE
Routing refactor

### PR DESCRIPTION
**Pull request by turbocrime**
beginning of handler/router refactor using connectrpc serverside concepts

this enables use of the handlercontext argument, simplifying internal routing, type checking, etc

### CODE REVIEW

**Comment by turbocrime** [comment]
also contains some docs for changes made in #255, docs might get split out

**Comment by turbocrime** [comment]
okay this kinda got away from me but i pulled it back a bit. this is deletes the old services router, and drops in new routers targeting serviceimpl and methodimpl types. no more type guards

all of the tests are broken but i wanted a sensible wip

File: packages/transport/src/chrome-runtime/adapter.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    //type AnyBiDiStreamingFn = (r: AsyncIterable<AnyMessage>, o?: CallOptions) => AsyncIterable<AnyMessage>;
+
+    switch (kind) {
+      case MethodKind.Unary: {
+        const res = (handlerMethod as AnyUnaryFn)(unpackedReq, opt);
+        return res.then(r => Any.pack(r).toJson({ typeRegistry }));
+      }
+      case MethodKind.ServerStreaming: {
+        const res = (handlerMethod as AnyServerStreamingFn)(unpackedReq, opt);
+        return Promise.resolve(iterableToStream(res).pipeThrough(new MessageToJson(typeRegistry)));
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/adapter.ts:80]: You can convert this return function to an async function and not need the Promise.resolve.

File: packages/router/src/ctx.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { Services } from '@penumbra-zone/services';
+
+import { localExtStorage, sessionExtStorage } from '@penumbra-zone/storage';
+import { AuthorizeRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/custody/v1alpha1/custody_pb';
+
+const services = new Services();
+void services.initialize();
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/ctx.ts:7]: wallet services and queriers provided to the routers as context. so, context creation needs a handle on it. this is also where init currently happens in main

File: packages/router/src/ctx.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import { createContextKey } from '@connectrpc/connect';
+
+export const servicesCtx = createContextKey(services);
+export const localCtx = createContextKey(localExtStorage);
+export const sessionCtx = createContextKey(sessionExtStorage);
+
+import { sendPopupRequest, spawnDetachedPopup } from '@penumbra-zone/types/src/internal-msg/popup';
+import { TxApproval } from '@penumbra-zone/types/src/internal-msg/tx-approval';
+
+const getTxApproval = async (req: AuthorizeRequest): Promise<void> => {
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/ctx.ts:18]: would use similar flow for, say, connected site approval.

goal is to split everything that is conceptually distinct from a serviceimpl/methodimpl out of the implementation.

think of this as an "approval provider"

File: packages/router/src/grpc/custody/index.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { CustodyProtocolService } from '@buf/penumbra-zone_penumbra.connectrpc_es/penumbra/custody/v1alpha1/custody_connect';
+import { authorize } from './authorize';
+import { ServiceImpl } from '@connectrpc/connect';
+
+export const custodyImpl: Partial<ServiceImpl<typeof CustodyProtocolService>> = {
+  authorize,
+} satisfies Omit<
+  ServiceImpl<typeof CustodyProtocolService>,
+  'exportFullViewingKey' | 'confirmAddress'
+>;
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/custody/index.ts:10]: yes. the omit forces the developer to list everything not implemented. sorry, had a doc on this but pulled it for some reason...

File: packages/router/src/ctx.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import { AuthorizeRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/custody/v1alpha1/custody_pb';
+
+const services = new Services();
+void services.initialize();
+
+import { createContextKey } from '@connectrpc/connect';
+
+export const servicesCtx = createContextKey(services);
+export const localCtx = createContextKey(localExtStorage);
+export const sessionCtx = createContextKey(sessionExtStorage);
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/ctx.ts:13]: conceptually, context should be the source of all information that isnt present in the request

we can dynamically configure client-specific, service-specific, request-specific contexts

File: packages/router/src/proxy.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  <S extends ServiceType>(
+    st: S,
+    remote: PromiseClient<S>,
+    overrider: ServiceOverrideFn = noOverride,
+  ) =>
+  ({ service }: ConnectRouter) => {
+    const passthru = createProxyService(remote);
+    const wrappedWithOverrides = overrider(passthru);
+    service(st, wrappedWithOverrides);
+  };
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/proxy.ts:66]: yeah this was an early sketch that turned out to be unnecessary because adapter also had similar behavior. its not imported anywhere.

File: packages/transport/src/chrome-runtime/adapter.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    if (!handlerMethod)
+      throw new ConnectError(
+        `No handler for ${reqType.typeName} of ${serviceType.typeName}`,
+        ConnectErrorCode.NotFound,
+      );
+
+    type AnyUnaryFn = (r: AnyMessage, o?: CallOptions) => Promise<AnyMessage>;
+    type AnyServerStreamingFn = (r: AnyMessage, o?: CallOptions) => AsyncIterable<AnyMessage>;
+    //type AnyClientStreamingFn = (r: AsyncIterable<AnyMessage>, o?: CallOptions) => Promise<AnyMessage>;
+    //type AnyBiDiStreamingFn = (r: AsyncIterable<AnyMessage>, o?: CallOptions) => AsyncIterable<AnyMessage>;
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/transport/src/chrome-runtime/adapter.ts:71]: these exist because the types i wanted from within connectrpc are not exported. wasnt sure where they might belong

File: packages/transport/src/chrome-runtime/adapter.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @param serviceType ServiceType of the service
+ * @param impl Object containing an implementation of the service
+ * @param remote Client to a remote service, for fallback/proxy
+ * @returns (req: JsonValue) => Promise<JsonValue> | ReadableStream<JsonValue>
+ */
+export const adaptServiceImpl = <S extends ServiceType>(
+  serviceType: S,
+  impl: Partial<ServiceImpl<S>>,
+  //ctx?: ContextValues,
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/transport/src/chrome-runtime/adapter.ts:35]: this could provide a "global context" i cut it out because i was uncertain about wide reuse of a context object. likely also a problem at service-worker.ts

 should probably use ContextInit type

File: packages/router/src/grpc/view-protocol-server/tx-planner.ts [hunk 9]
```diff
@@ -36,29 +29,17 @@ export const handleTxPlannerReq = async (
...
-    if (!o.value || !o.address) continue;
-    planner.output(o.value, o.address);
-  }
+  for (const { value, address } of req.outputs)
+    if (value && address) planner.output(value, address);
 
-  for (const w of req.ics20Withdrawals) {
-    planner.ics20Withdrawal(w);
-  }
+  for (const w of req.ics20Withdrawals) planner.ics20Withdrawal(w);
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/tx-planner.ts:39]: linter rule

File: packages/router/src/grpc/view-protocol-server/nullifier-status.ts [hunk 10]
```diff
@@ -1,47 +1,24 @@
...
+> = async (req, ctx) => {
+  const services = ctx.values.get(servicesCtx);
   await assertWalletIdMatches(req.walletId);
   if (!req.nullifier) throw new Error('No nullifier passed');
-
   const { indexedDb } = await services.getWalletServices();
-  const spent = await noteOrSwapSpent(indexedDb, req.nullifier);
+  const swapByNullifier = await indexedDb.getSwapByNullifier(req.nullifier);
+  const noteByNullifier = await indexedDb.getNoteByNullifier(req.nullifier);
+  const spent = Boolean(noteByNullifier?.heightSpent) || Boolean(swapByNullifier?.heightClaimed);
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/nullifier-status.ts:19]: i think it is pretty widely acknowledged that 0 is falsy

File: packages/router/src/grpc/view-protocol-server/index.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import { nullifierStatus } from './nullifier-status';
+import { status } from './status';
+import { statusStream } from './status-stream';
+import { transactionInfoByHash } from './tx-info-by-hash';
+import { transactionInfo } from './tx-info';
+import { transactionPlanner } from './tx-planner';
+import { walletId } from './wallet-id';
+import { witnessAndBuild } from './witness-build';
+
+export const viewImpl: Omit<
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/index.ts:23]: would like to consider dropping all the method impls into this one file.

alternatively, do something that scopes the method type and ctx keys to every method file automatic, so each method file is just a bare handler.

File: packages/router/src/grpc/view-protocol-server/index-by-address.ts [hunk 12]
```diff
@@ -1,24 +1,19 @@
...
+) => {
+  const local = ctx.values.get(localCtx);
+  const wallets = await local.get('wallets');
   if (!req.address) throw new Error('no address given in request');
 
   const address = bech32Address(req.address);
   const addressIndex = isControlledAddress(wallets[0]!.fullViewingKey, address);
 
-  return addressIndex ? new IndexByAddressResponse({ addressIndex }) : new IndexByAddressResponse();
+  return addressIndex != null ? { addressIndex } : {};
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/index-by-address.ts:18]: there is a distinction between a null index and a zero index, but just testing index boolean doesnt make that distinction

`null != undefined` and `null != null` and `undefined != undefined` are all false iirc

File: packages/transport/src/chrome-runtime/adapter.ts (hunk 1 above)
Comment (turbocrime) [reply packages/transport/src/chrome-runtime/adapter.ts:80]: the pipeTo creates a promise actually iirc so i will just drop the Promise.resolve

File: packages/transport/src/chrome-runtime/adapter.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (grod220) [reply packages/transport/src/chrome-runtime/adapter.ts:71]: They don't have to be exported, just moved to the top of the file and not in a function

File: packages/router/src/ctx.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/ctx.ts:7]: oh right, i moved things around a bit and confused this. thanks

File: packages/router/src/grpc/view-protocol-server/index-by-address.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/index-by-address.ts:14]: ah, i was assuming addressIndex was a plain number for some reason

File: packages/router/src/grpc/view-protocol-server/nullifier-status.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/nullifier-status.ts:16]: what is the appropriate behavior if a spend height of zero does appear on the chain?

[the remark this answers is not in the document]
Comment (hdevalence) [reply packages/router/src/grpc/view-protocol-server/nullifier-status.ts]: It's impossible. Block heights start at 1.  Technically speaking, there is not actually a genesis block. There's only a genesis _app state_, which is a state snapshot (rather than a block, which is a kind of state delta). We create a "phantom" compact block with height 0 to allow clients to scan the genesis state snapshot as if it were a delta from the empty state, but it is not a real block.

File: packages/router/src/grpc/view-protocol-server/tx-planner.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/tx-planner.ts]: i've restored the brackets but i think we should enforce code style via linter rules

File: packages/router/src/grpc/view-protocol-server/nullifier-status.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/nullifier-status.ts]: i've added a comment explaining heights. this is one of the handlers that need to be refactored to stream properly so this code will be revisited again pretty soon

File: packages/router/src/ctx.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/ctx.ts]: we don't have to expose it globally. right now the adapter/router doesn't discriminate what methods receive what context, but that feature is available with this refactor, and should eventually be used.

approval is an implementation detail of the extension hosting the service, not an implementation detail of the method requiring the approval data. if we are designing these services to be reusable by third parties, we should abstract all details of the service's 'host' into context.

this kind of abstraction is also important if we want to eventually support things like approval when proxying page-described services.

File: packages/router/src/ctx.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (turbocrime) [reply packages/router/src/ctx.ts]: if we are designing these services to be reusable by third parties, we should abstract all details of the service's 'host' into context.

a well-organized context allows the service implementation to be modular, and provides third-party developers a clear description of what the service requires, and an interface to provide those requirements.

**Comment by turbocrime** [comment]
significantly changed adapter.

adapter now targets a ConnectRouter, rather than directly calling implementations. this should be closer to the behavior of a typical connectrpc server.

a single router can represent multiple services, so it should be possible now to simplify connection management and the client, and use a single transport for multiple clients.

File: packages/router/src/grpc/view-protocol-server/nullifier-status.ts [hunk 13]
```diff
@@ -1,91 +1,58 @@
...
+        const swap = SwapRecord.fromJson(swapJson);
+        return Boolean(swap.heightClaimed) && nullifier.equals(swap.nullifier);
+      }),
+      watchStream(noteStream, ({ value: noteJson }) => {
+        const note = SpendableNoteRecord.fromJson(noteJson);
+        return Boolean(note.heightSpent) && nullifier.equals(note.nullifier);
+      }),
+    ]);
+
+    return eventuallySpent.then(() => ({ spent: true }));
```
Comment (turbocrime) [inline packages/router/src/grpc/view-protocol-server/nullifier-status.ts:54]: @grod220 used Promise.race to independently iterate, rather than attempting to merge asynciterators

File: packages/router/src/grpc/clients.ts
Comment (turbocrime) [inline packages/router/src/grpc/clients.ts:1]: now using tendermint querier from services. is there a reason this client was present?

File: apps/extension/src/service-worker.ts [hunk 14]
```diff
@@ -1,111 +1,97 @@
...
+  (
+    message: unknown,
+    _: chrome.runtime.MessageSender,
+    sendResponse: (response: unknown) => void,
+  ) => {
+    if (!isStdRequest(message)) return;
+    stdRouter(message, sendResponse, services);
+    return true;
+  },
+);
```
Comment (turbocrime) [inline apps/extension/src/service-worker.ts:66]: not much remaining that still uses stdClient. would like popup control to be independent, and 'sync' is a thing that probably doesn't need a specific request. those bits feel out of scope for this pr

File: packages/router/src/grpc/view-protocol-server/nullifier-status.ts
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/nullifier-status.ts:1]: This is a good candidate for its own PR. In the future, let's keep some of these side issues out of large PRs like this one.

File: packages/transport/src/chrome-runtime/adapter.ts [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+// from createRouterTransport
+const forceRouterOptions = {
+  connect: true,
+  grpc: false,
+  grpcWeb: false,
+};
+
+// from createRouterTransport
+const forceTransportOptions = {
+  httpClient: null as never,
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/adapter.ts:56]: do we need `as never`?

File: packages/transport/src/chrome-runtime/adapter.ts [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+  // TODO: alternatively, we could have the channelClient provide a requestPath
+  const I_MethodType = new Map<string, MethodType>(
+    router.handlers.map(({ method, service }) => [
+      method.I.typeName,
+      { ...method, service: { typeName: service.typeName } },
+    ]),
+  );
+
+  // TODO: interceptors?
```
Comment (grod220) [inline packages/transport/src/chrome-runtime/adapter.ts:153]: What are interceptors? Can you make a more descriptive TODO if it's for later PRs?

File: packages/transport/src/create.ts [hunk 17]
```diff
@@ -142,14 +119,17 @@ export const createChannelTransport = ({
...
         } catch (error) {
           const err = ConnectError.from(error);
-          // throwing from a listener is pointless. store the first error, log
+          // some fundamental disaster has occurred, and now the transport is
+          // broken forever. we have no specific request to reject, but throwing
+          // from a listener is a dead end. so we store it to throw at every
+          // subsequent interaction.
           listenerError ??= err;
-          console.error('Error in transport listener', err);
+          console.warn('Transport failed', err);
```
Comment (grod220) [inline packages/transport/src/create.ts:127]: If it's a fundamental disaster kind of condition, think console.error is probably appropriate

File: packages/transport/src/types/channel-names.ts [hunk 18]
```diff
@@ -37,7 +37,7 @@ export enum ChannelClientLabel {
 }
 
 export enum ChannelSubLabel {
-  ClientStream = 'ClientStream',
+  //ClientStream = 'ClientStream',
```
Comment (grod220) [inline packages/transport/src/types/channel-names.ts:40]: Delete?

**Comment by grod220** [comment]
Also, does this refactor support the case where the view server makes a request to the custody server? We will need this to implement https://github.com/penumbra-zone/web/issues/211 and fulfill https://github.com/penumbra-zone/web/issues/324

**Comment by turbocrime** [comment]
> Also, does this refactor support the case where the view server makes a request to the custody server? We will need this to implement #211 and fulfill #324

haven't tried this specifically yet but should be simple as putting a client into context, putting the necessary information into context, or calling the handler directly and forwarding context.

File: packages/router/src/grpc/view-protocol-server/ephemeral-address.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+  const {
+    viewServer: { fullViewingKey },
+  } = await services.getWalletServices();
+
+  let address;
+  try {
+    address = wasm.getEphemeralByIndex(fullViewingKey, req.addressIndex?.account ?? 0);
+  } catch (wasmErr) {
+    throw new ConnectError('WASM failed to generate address', Code.Internal);
+  }
```
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/ephemeral-address.ts:19]: Can we remove this let+try/catch+throw pattern in this PR (think there are ~10 instances) now that it gets propagated somewhere else?

File: packages/query/src/queriers/tendermint.ts [hunk 20]
```diff
@@ -16,4 +19,12 @@ export class TendermintQuerier implements TendermintQuerierInterface {
     const res = await this.client.getStatus(req);
     return res.syncInfo!.latestBlockHeight;
   }
+
+  async broadcastTx(params: Uint8Array) {
```
Comment (grod220) [inline packages/query/src/queriers/tendermint.ts:23]: Think we've had a new method added here in a recent PR merge

File: packages/router/src/grpc/view-protocol-server/wallet-id.ts [hunk 21]
```diff
@@ -1,22 +1,17 @@
...
-  const wallets = await localExtStorage.get('wallets');
+import { WalletId } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/keys/v1alpha1/keys_pb';
 
-  if (!wallets.length) return new WalletIdResponse();
+import { ConnectError, Code } from '@connectrpc/connect';
 
-  return new WalletIdResponse({
-    walletId: new WalletId({ inner: stringToUint8Array(wallets[0]!.id) }),
-  });
+// TODO: there is never data in this request. should a wallet index be available in context?
```
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/wallet-id.ts:10]: Don't think there is such a thing as wallet index in the rpc defintion

File: packages/transport/src/chrome-runtime/adapter.ts (hunk 15 above)
Comment (grod220) [reply packages/transport/src/chrome-runtime/adapter.ts:56]: I'd say just add a code comment next to the line that explains it

**Review by grod220 — APPROVED** [review]

**Comment by TalDerei** [comment]
Once we resolve the conflicts with main, LGTM!

**Review by TalDerei — APPROVED** [review]

File: packages/query/src/queriers/tendermint.ts (hunk 20 above)
Comment (turbocrime) [reply packages/query/src/queriers/tendermint.ts:23]: rebased

File: packages/router/src/grpc/view-protocol-server/wallet-id.ts (hunk 21 above)
Comment (turbocrime) [reply packages/router/src/grpc/view-protocol-server/wallet-id.ts:10]: ok. there are a few places in the code that seem to require a wallet index to access storage, often defaulting to 0. seems like a separate issue
